kv/kvserver: TestReplicateRestartAfterTruncationWithRemoveAndReAdd failed

kv/kvserver.TestReplicateRestartAfterTruncationWithRemoveAndReAdd [failed](https://mesolite.cluster.engflow.com/invocations/default/7b57c400-b7eb-48c0-a420-897f31c33988?testReportRun=8&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [b07e9c2cc91be05a90b1c0f0665da0da4cde2e7a](https://github.com/cockroachdb/cockroach/commits/b07e9c2cc91be05a90b1c0f0665da0da4cde2e7a):


```
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
          |   | 	pkg/rpc/context.go:181
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
          |   | 	pkg/rpc/auth.go:97
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/rpc.gatewayRequestRecoveryInterceptor
          |   | 	pkg/rpc/metrics.go:554
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/server.newGRPCServer.NewRequestMetricsInterceptor.func4
          |   | 	pkg/rpc/metrics.go:472
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
          |   | 	pkg/rpc/context.go:138
          |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          |   | 	pkg/util/stop/stopper.go:351
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
          |   | 	pkg/rpc/context.go:136
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func5
          |   | 	pkg/rpc/context.go:878
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func2
          |   | 	pkg/rpc/context.go:777
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewContext.ClientInterceptor.func9
          |   | 	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:185
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.getChainUnaryInvoker.func4
          |   | 	pkg/rpc/context.go:962
          |   | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func3
          |   | 	pkg/rpc/context.go:847
          |   | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          |   | 	pkg/rpc/context.go:970
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          |   | 	pkg/kv/kvclient/kvcoord/transport.go:211
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          |   | 	pkg/kv/kvclient/kvcoord/transport.go:189
          | Wraps: (4) change replicas of r80 failed
          | Wraps: (5) range r80:/System{\x00aaa-testing-/NodeLiveness} [(n1,s1):1, (n3,s3):3, (n2,s2):4LEARNER, next=5, gen=9, sticky=9223372036.854775807,2147483647] cannot make progress with proposed changes add=[(n2,s2):4LEARNER] del=[] based on live replicas [(n1,s1):1]
          |   |
          |   | (opaque error leaf)
          |   | type name: github.com/cockroachdb/cockroach/pkg/kv/kvserver/*kvserver.quorumError
          | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errbase.opaqueWrapper (4) *errutil.withPrefix (5) *errbase.opaqueLeaf
        Error types: (1) *barriers.barrierErr
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicateRestartAfterTruncationWithRemoveAndReAdd1062943213
--- FAIL: TestReplicateRestartAfterTruncationWithRemoveAndReAdd (11.50s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=8</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateRestartAfterTruncationWithRemoveAndReAdd.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56412

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -36,7 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/livenesspb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/rditer"
-	raft "github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft"
 	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/rpc"
@@ -2294,6 +2294,12 @@ func runReplicateRestartAfterTruncation(t *testing.T, removeBeforeTruncateAndReA
 					StickyVFSRegistry: fs.NewStickyRegistry(),
 					WallClock:         manualClock,
 				},
+				// In this test, under duress, we can occasionally fail enough
+				// heartbeats to exceed the 10s timeout in execChangeReplicasTxn to
+				// verify liveness of a post-change quorum.
+				//
+				// See: https://github.com/cockroachdb/cockroach/issues/156689
+				Store: &kvserver.StoreTestingKnobs{AllowDangerousReplicationChanges: true},
 			},
 		}
 	}

